src/t_cose_mac_verify.c

```diff
@@ -0,0 +1,335 @@
+     *   local vars                                    20          16
+     *   TOTAL                                         20          16
+     */
+    uint64_t uTag;
+    uint32_t item_tag_index = 0;
+    int returned_tag_index;
+
+    /* The 0th tag is the only one that might identify the type of the
+     * CBOR we are trying to decode so it is handled special.
+     */
+    uTag = QCBORDecode_GetNthTagOfLast(decode_context, item_tag_index);
+    item_tag_index++;
+    if(me->option_flags & T_COSE_OPT_TAG_REQUIRED) {
+        /* The protocol that is using COSE says the input CBOR must
+         * be a COSE tag.
+         */
+        if(uTag != CBOR_TAG_COSE_SIGN1) {
+            return T_COSE_ERR_INCORRECTLY_TAGGED;
+        }
+    }
+    if(me->option_flags & T_COSE_OPT_TAG_PROHIBITED) {
+        /* The protocol that is using COSE says the input CBOR must
+         * not be a COSE tag.
+         */
+        if(uTag == CBOR_TAG_COSE_SIGN1) {
+            return T_COSE_ERR_INCORRECTLY_TAGGED;
+        }
+    }
+    /* If the protocol using COSE doesn't say one way or another about the
+     * tag, then either is OK.
+     */
+
+
+    /* Initialize auTags, the returned tags, to CBOR_TAG_INVALID64 */
+#if CBOR_TAG_INVALID64 != 0xffffffffffffffff
+#error Initializing return tags array
+#endif
+    memset(me->auTags, 0xff, sizeof(me->auTags));
+
+    returned_tag_index = 0;
+
+    if(uTag != CBOR_TAG_COSE_SIGN1) {
+        /* Never return the tag that this code is about to process. Note
+         * that you can sign a COSE_SIGN1 recursively. This only takes out
+         * the one tag layer that is processed here.
+         */
+        me->auTags[returned_tag_index] = uTag;
+        returned_tag_index++;
+    }
+
+    while(1) {
+        uTag = QCBORDecode_GetNthTagOfLast(decode_context, item_tag_index);
+        item_tag_index++;
+        if(uTag == CBOR_TAG_INVALID64) {
+            break;
+        }
+        if(returned_tag_index > T_COSE_MAX_TAGS_TO_RETURN) {
+            return T_COSE_ERR_TOO_MANY_TAGS;
+        }
+        me->auTags[returned_tag_index] = uTag;
+        returned_tag_index++;
+    }
+
+    return T_COSE_SUCCESS;
+}
+
+#ifndef T_COSE_DISABLE_SHORT_CIRCUIT_SIGN
+/**
+ *  \brief Verify a short-circuit tag
+ *
+ * \param[in] cose_alg_id  Algorithm ID. This is used only to make
+ *                         the short-circuit signature the same size as the
+ *                         real tag would be for the particular algorithm.
+ * \param[in] header       The Header of COSE_Mac0.
+ * \param[in] payload      The payload of COSE_Mac0
+ * \param[in] tag          Pointer and length of tag to be verified
+ *
+ * \return This returns one of the error codes defined by \ref
+ *         t_cose_err_t.
+ *
+ * See short_circuit_tag() in t_cose_mac_sign.c for description of
+ * the short-circuit tag.
+ */
+static inline enum t_cose_err_t
+short_circuit_verify(int32_t               cose_alg_id,
```

Comment:
I think the whole short-circuit thing can be eliminated for MAC. It was there for signing because the pubic key crypto is harder to integrate and test with and because the keys are harder to generate and such. This is much less true of HMAC, so I think it can be removed.